File: JavaSe/src/main/java/com/arithmetic/recursion/RecursionTest.java
package com.arithmetic.recursion;

/**
 * @author ddone
 * @date 2020/7/19-19:32
 */
public class RecursionTest {
    public static void main(String[] args) {
        int num = 50000;
        System.out.println(tailRecSum(num,0));
//        System.out.println(recSum(num));
    }

    public static Integer recSum(int num){
        if(num==1){
           return 1;
        }else {
            return recSum(num - 1)+num;
        }
    }
    public static Integer tailRecSum(int num,int sum){
        if(num ==0){
            return sum ;
        }else {
            return tailRecSum(num - 1, sum+num);
        }
    }
}

File: Scala_Spark/src/main/scala/com/ddone/RecursionTest.scala
package com.ddone

/**
 **
@author ddone
 **
 * @date 2020/7/19-19:46
 *
 */
object RecursionTest {
  def main(args: Array[String]): Unit = {
    println(tailRecSum(10000000,0))
    println(recSum(10000))
  }

  def recSum(num:Int):Int={
    if (num ==1 ){
       1;
    }else{
      num+recSum(num-1)
    }
  }

  def tailRecSum(num:Int,sum:Int): Int ={
    if (num==0){
      sum
    }else{
      tailRecSum(num-1,sum+num)
    }
  }
}
